commands/command_fetch.go

```diff
@@ -21,8 +22,65 @@ var (
 	fetchAllArg    bool
 	fetchPruneArg  bool
 	fetchDryRunArg bool
+	fetchJsonArg   bool
 )
 
+type FetchWatcher struct {
+	transfers        []*tq.Transfer
+	virtuallyFetched map[string]bool
+}
+
+func newFetchWatcher() *FetchWatcher {
+	ret := &FetchWatcher{
+		transfers:        nil,
+		virtuallyFetched: nil,
+	}
+	if fetchJsonArg {
+		ret.transfers = make([]*tq.Transfer, 0)
+	}
+	if fetchDryRunArg {
+		ret.virtuallyFetched = make(map[string]bool)
+	}
+	return ret
+}
+
+func (d *FetchWatcher) registerTransfer(t *tq.Transfer) {
+	if d.transfers != nil {
+		d.transfers = append(d.transfers, t)
+	}
+	if d.virtuallyFetched != nil {
+		d.virtuallyFetched[t.Oid] = true
+	}
+	if fetchDryRunArg {
```

Comment:
I hemmed and hawed a bit over this, because `append()` will accept a `nil` slice and just allocate a backing array, so we could drop the `make([]*tq.Transfer, 0)` in the `newFetchWatcher()` function above, and then do something like:
```go
	if fetchJsonArg {
		d.transfers = append(d.transfers, t)
	}
	if fetchDryRunArg {
		d.virtuallyFetched[t.Oid] = true
		printProgress("%s %s => %s", tr.Tr.Get("fetch"), t.Oid, t.Name)
	}
```

However, since in your next PR #5975 you're adding another flag whose state affects both functions, I think it's probably more elegant to leave it as you've written it.

Reply:
Fair point, I'll mull it over when I get back to that PR, maybe I can indeed move choosing the behaviour from the constructor to this function. For now, I left it like it was.